atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java

```diff
@@ -1891,6 +1898,108 @@ public void transactionStillCommittedEvenIfCallbackThrows() {
         });
     }
 
+    @Test
+    public void cannotCompleteFutureAfterTransactionCommit() {
+        SettableFuture<Object> blocker = SettableFuture.create();
+        ListenableFuture<Map<Cell, byte[]>> getFuture = txManager.runTaskThrowOnConflict(txn -> {
+            txn.put(TABLE, ImmutableMap.of(TEST_CELL, PtBytes.toBytes("jeremy")));
+            return Futures.transformAsync(
+                    blocker,
+                    _unused -> txn.getAsync(TABLE, ImmutableSet.of(TEST_CELL)),
```

Comment:
This is testing something subtly different if I am not mistaken: this is testing that you cannot complete a transaction, then call a method on that transaction. I think you ideally want a test where you start `getAsync` well within the bounds of the transaction, but that you return that (unfinished) future and use that outside. I don't think this matters except in SerializableTransactions, and I appreciate that this is probably hard to test, but that's the behaviour here that was at fault before, I believe.

Reply:
You're right: yeah the getAsync only happens after the transaction has completed here. I added one that handles usage of a future that began within (and completed within) but is used outside.

There isn't really a great way of injecting this, though, without considerably rewiring the SnapshotTransaction :(